### ISSUE TITLE
Support passing custom parameters to Neutrino via the CLI

### ISSUE THREAD

**Issue by edmorley** [mozilla-neutrino/neutrino-dev#165]
A typical way of passing options to commands would be to define the package.json script entry like so:
`NODE_ENV=production FOO=bar webpack -p --progress`

Unfortunately on Windows environment variables must instead be set like `set NODE_ENV=production`, so to write a cross-platform compatible package.json, previously the recommended solution was to use cross-env:
`cross-env NODE_ENV=production FOO=bar webpack -p --progress`

However this adds yet another dependency and boilerplate to package.json.

webpack 2 now supports an `--env` flag for passing of arbitrary options, so the above could be replaced with:
`webpack -p --env.production --env.foo=bar --progress`

...which is now the recommended way to pass options in a cross-platform compatible way, according to the cross-env author. See the second half of this guide (search for "cross-env"):
https://blog.flennik.com/the-fine-art-of-the-webpack-2-config-dc4d19d7f172#.297u8iuz1

However currently Neutrino doesn't support passing arbitrary arguments (at least according to the usage text and/or CLI docs).

I don't have a strong preference as to what format the flags should take, but I imagine it might be more intuitive if Neutrino just copied webpack 2's syntax to save having to remember another.

Webpack docs:
https://webpack.js.org/guides/environment-variables/
https://webpack.js.org/configuration/configuration-types/

**Comment by edmorley** [comment]
So unfortunately Neutrino already supports an `--env` parameter, however it's used to set the value of `NODE_ENV`:
https://github.com/mozilla-neutrino/neutrino-dev/blob/3f56635381131369279ea54c931c8277ef39aca5/packages/neutrino/bin/neutrino#L28-L32
https://github.com/mozilla-neutrino/neutrino-dev/blob/3f56635381131369279ea54c931c8277ef39aca5/packages/neutrino/bin/neutrino#L66

I still think it would be preferable to match webpack's parameter scheme, though this would mean having to support the old `--env=production` form as well as the new `--env.foo=bar` form.

For reference, webpack's env parameter handling can be found here:
https://github.com/webpack/webpack/blob/38f9686a5d11a15d5d14c8f9008093da0f5bb5b3/bin/config-yargs.js#L23-L26
https://github.com/webpack/webpack/blob/38f9686a5d11a15d5d14c8f9008093da0f5bb5b3/bin/convert-argv.js#L100-L103

**Comment by edmorley** [comment]
> I still think it would be preferable to match webpack's parameter scheme, though this would mean having to support the old --env=production form as well as the new --env.foo=bar form.

@eliperelman what are your thoughts on approach for this? Keen to get this fixed since it would make the cross platform concerns of mozilla/treeherder#2309 go away :-)

**Comment by eliperelman** [comment]
We could break the `--env`  flag in v6, and allow switching the `NODE_ENV` via something like `--env.NODE_ENV production`.

**Comment by eliperelman** [comment]
So, I'm starting to lean away from doing this. We can already override the env via `--env`.

We have two issues here:

1. Setting environment variables that middleware or Neutrino may consume
2. Setting configuration values from the command line


#### Setting environment variables that middleware or Neutrino may consume

Currently the environment variables that are used are `NODE_ENV`, `PORT`, `HOST`, `HTTPS`, and `CI`.

- Should we get away from most of these environment variables and use options instead?
- Why should Neutrino facilitate a mechanism for **setting** these environment variables in a cross-platform manner? This seems like an external concern.

#### Setting configuration values from the command line

I feel like allow this would complicate options. What is the value add of allowing this vs. done from an override file?

**Comment by edmorley** [comment]
> I feel like allow this would complicate options. What is the value add of allowing this vs. done from an override file?

Because cross platform environment variables in package.json currently requires annoying boilerplate (the cross-env package) and the author of cross-env recommends using the new webpack way of passing parameters on the CLI, which Neutrino blocks using.

**Comment by edmorley** [comment]
As an example, this is not cross-platform compatible:
```json
  "scripts": {
    "start": "neutrino start --presets ./neutrino-custom/development.js",
    "start:stage": "BACKEND_DOMAIN=https://treeherder.allizom.org neutrino start --presets ./neutrino-custom/development.js",
    "start:prod": "BACKEND_DOMAIN=https://treeherder.mozilla.org neutrino start --presets ./neutrino-custom/development.js",
  }
```

Whereas this would be:
```json
  "scripts": {
    "start": "neutrino start --presets ./neutrino-custom/development.js",
    "start:stage": "neutrino start --env.BACKEND_DOMAIN=https://treeherder.allizom.org --presets ./neutrino-custom/development.js",
    "start:prod": "neutrino start --env.BACKEND_DOMAIN=https://treeherder.mozilla.org --presets ./neutrino-custom/development.js",
  }
```

**Comment by eliperelman** [comment]
Resolved in #233. This means Ed's use case would be solved with:

```json
"scripts": {
  "start": "neutrino start",
  "start:stage": "neutrino start --options.env.BACKEND_DOMAIN https://treeherder.allizom.org",
  "start:prod": "neutrino start --options.env.BACKEND_DOMAIN https://treeherder.mozilla.org",
}
```

There is no more `--env` CLI option.

**Comment by edmorley** [comment]
Awesome, thank you!
